### ISSUE TITLE
[Feature] SelectorGenerator as standalone package

### ISSUE THREAD

**Issue by samdoeswork** [microsoft/playwright#13900]
Hi,
The selector-generator component which is injected as part of the codegen mode has significant utility outside of playwright.
I'd really like to use it stand-alone. I think there's a lot of projects which would like to also.

Would you consider publishing this independely, or providing a code-sample which uses just this alone?

Thanks!

**Comment by dgozman** [comment]
@samdoeswork Tell us more about your usecase? Where do you intend to use it? We don't have plans to provide a standalone package right now, but more details will help us prioritize it.

**Comment by samdoeswork** [comment]
Thanks @dgozman 

The selector-generator is so elegant. My use-case is somewhat niche: detecting when elements appear during manual browser-sessions.

For my case, I want to use your brilliant selector-generation engine to generate data on what a user does in browser.
The goal (for me) isn't to record and replay, it's to record and use that data.

- Imagine editing tutorial videos & having precise timestamps on when you did various actions so you can very easily edit the video, and/or automatically provide chapter markers to the user.
- Imagine automatic record-keeping for timesheet-bound employees or highly regulated industries

For my particular use case I want to automatically timestamp media, by correlating in-browser elements of interest with the time in the recording -- making these elements searchable, jump-to, bulk edit, etc.

There's a very long tail of times where it's useful to identify moments when you did something.

I realize it's niche compared to record & replay.

Another obvious use-case is de-coupling the record & replay: people may need to lightly customize one part or both.

**Comment by dgozman** [comment]
@samdoeswork These are interesting ideas, but seem to be out of scope for Playwright, at least right now. That said, I'll keep the issue open just in case. Meanwhile, feel free to fork/copy the selector generator, it is open source after all 😄

**Comment by AndyGleed** [comment]
> @samdoeswork These are interesting ideas, but seem to be out of scope for Playwright, at least right now. That said, I'll keep the issue open just in case. Meanwhile, feel free to fork/copy the selector generator, it is open source after all 😄

@dgozman  Hi, i have been trying to locate the selector generator to folk as I have some ideas I would like to try out with it, but I can't find it.  Could you please link me, thanks :)

**Comment by dgozman** [comment]
@AndyGleed [Here you go](https://github.com/microsoft/playwright/blob/main/packages/playwright-core/src/server/injected/selectorGenerator.ts).

**Comment by GeorgeFischhof** [comment]
Hi,
separate selector generator is good when you are not beginner in the code, and during the code writing you need a selector.
Without selector generator you have to use developer tools, and find out a good selector, which is time consuming, and breaks the flow of the development.
Also recorder creates actions, but when the engineer wants to check something, then creation of an expect statement would be good #14913

**Comment by GeorgeFischhof** [comment]
I tried the debug mode using the codegen UI and really an "Explore" button appeared. I think it should be visible by default. Even without debugging

**Comment by samdoeswork** [comment]
@dgozman thank you for your help so far!

Here's our plan. Does this sound correct?

If we import InjectedScript() class, we could theoretically use that to create a context in a browser extension to utilize and generate Playwright compatible selectors?

I'd much rather be Playwright compatible than write something similar for many reasons (including frankly that your implementation is better than what I could manage!)

**Comment by dgozman** [comment]
@samdoeswork That's right. You'll need to copy code from the `injected/` directory, instantiate `InjectedScript` and call `generateSelector()`.

**Comment by puranjayjain** [comment]
is there a small example on how to do it?

**Comment by puranjayjain** [comment]
I'm not able to understand the usage of injectedScript and how do i pass it can someone help me with an example? @dgozman or @samdoeswork please ?

**Comment by entrptaher** [comment]
Here is an example.

```js
const browser = await chromium.launch();
const context = await browser.newContext();
const page = await context.newPage();
await page.goto('http://example.com')

// find the element some other way
const h1 = await page.locator("xpath=//h1");
const handle = await h1.elementHandle();

// get the generated locator from that element
const generatedLocator = await page.evaluate((handle) => {
  return {
    selector: playwright.selector(handle),
    locator: playwright.generateLocator(handle),
  };
}, handle);

console.log(generatedLocator);
```

To get this, you will need to enable the console, here is how to do that.

```
PWDEBUG=console node index.mjs
```

**Comment by kolodny** [comment]
I had a similar need and ended up extracting Playwright's browser-side internals into a standalone package: [`playwright-injected`](https://github.com/kolodny/playwright-injected). It's a zero-dependency ES module with the InjectedScript, selector engine, locator generation, and ARIA utils that you can import and run in the browser — no Playwright install needed.

```js
import { InjectedScript, asLocator } from 'playwright-injected';

const injected = new InjectedScript(window, {
  isUnderTest: false,
  sdkLanguage: 'javascript',
  testIdAttributeName: 'data-testid',
  stableRafCount: 0,
  browserName: 'chromium',
  customEngines: [],
});

const el = document.querySelector('button');
const { selector, selectors } = injected.generateSelector(el);
// selector  => 'internal:role=button[name="Submit"i]'
// selectors => ['internal:role=button[name="Submit"i]', '#submit-btn', ...]

asLocator('javascript', selector);
// => "getByRole('button', { name: 'Submit' })"
```
